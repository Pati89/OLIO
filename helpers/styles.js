// here we can add theme, fontSizes or spacing sizes to use same all over the page

export const breakpoints = {
  sm: 640,
  md: 896,
  lg: 1092,
};

export const media = {
  /* small size and up */
  sm: `@media only screen and (min-width: ${breakpoints.xs}px)`,
  /* medium size and up */
  md: `@media only screen and (min-width: ${breakpoints.sm}px)`,
  /* large size and up */
  lg: `@media (min-width: ${breakpoints.md}px)`,
};
