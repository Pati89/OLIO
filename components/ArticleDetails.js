import PropTypes from "prop-types";
import Image from "next/image";
import Link from "next/link";
import styled from "@emotion/styled";
import { media } from "../helpers/styles";
import { getByText } from "@testing-library/react";

const S_div_container = styled.div`
  width: 70%;
  height: 100%;
  text-align: left;
  color: inherit;
  text-decoration: none;
  overflow: hidden;
  cursor: pointer;
  background-color: white;

  ${media.md} {
    width: 80%;
  }
`;

const S_div_image = styled.div`
  background-color: lightgray;
  position: relative;
  width: 100%;
  height: 200px;
  max-height: 300px;

  ${media.md} {
    height: 400px;
    max-height: 400px;
  }
`;

const S_div_card_content = styled.div`
  width: 100%;
  padding: 20px;

  ${media.md} {
    width: auto;
  }
`;

const S_h1_title = styled.h1`
  color: inherit;
  font-weight: bold;
  margin-top: 0;
`;

const S_div_user_container = styled.div`
  display: flex;
  justify-content: space-between;
  align-items: center;
`;

const S_div_avatar = styled.div`
  display: flex;
  position: relative;
  width: 30px;
  height: 30px;
  border-radius: 15px;
  overflow: hidden;
`;

const S_div_user = styled.div`
  display: flex;
  align-items: center;
`;

const S_div_status = styled.div`
  display: flex;
  align-items: center;
  background-color: ${({ status }) => (status === "active" ? "green" : "grey")};
  color: white;
  padding: 5px;
  border-radius: 10px;
  font-size: 0.75rem;
`;

const S_div_divider = styled.div`
  width: calc(100% - 40px);
  margin: 0 20px;
  height: 1px;
  border-top: 1px solid lightgrey;
`;

const S_div_description = styled.div`
  padding: 0 20px 20px;
`;

const S_div_description_content = styled.div`
  white-space: pre-wrap;
`;

const S_p_description_title = styled.p`
  font-weight: bold;
`;

const S_span_user = styled.span`
  margin-left: ${({ avatarSrc }) => (avatarSrc ? "10px" : 0)};
`;

const S_a_back = styled.a`
  display: flex;
  align-items: center;
  width: 70%;
  background-color: rgb(240, 238, 238);
  padding: 10px 0;

  ${media.md} {
    width: 80%;
  }
`;

const S_i_arrow_back = styled.i`
  border: solid black;
  border-width: 0 2px 2px 0;
  display: inline-block;
  padding: 3px;
  transform: rotate(135deg);
  margin-right: 10px;
`;

const Article = (props) => {
  const { article } = props;

  const imageSrc = article?.images[0]?.files?.large;
  const articleTitle = article?.title;
  const avatarSrc = article?.user?.current_avatar?.small;
  const user = article?.user?.first_name;
  const status = article?.status;
  const description = article?.description || "No description";

  return (
    <>
      <Link href="/" passHref>
        <S_a_back>
          <S_i_arrow_back /> Back
        </S_a_back>
      </Link>
      <S_div_container data-testid="articleDetails">
        {imageSrc && (
          <S_div_image>
            <Image
              className="image"
              src={imageSrc}
              alt={articleTitle}
              layout="fill"
            />
          </S_div_image>
        )}
        <S_div_card_content>
          <S_h1_title>{articleTitle}</S_h1_title>
          <S_div_user_container>
            <S_div_user>
              {avatarSrc && (
                <S_div_avatar data-testid="avatar">
                  <Image src={avatarSrc} alt={user} layout="fill" />
                </S_div_avatar>
              )}
              <S_span_user avatarSrc={avatarSrc}>{user}</S_span_user>
            </S_div_user>
            <S_div_status status={status} data-testid="status">
              {status}
            </S_div_status>
          </S_div_user_container>
        </S_div_card_content>
        <S_div_divider />
        <S_div_description>
          <S_p_description_title>Description:</S_p_description_title>
          <S_div_description_content
            data-testid="description"
            dangerouslySetInnerHTML={{
              __html: `${description}`,
            }}
          />
        </S_div_description>
      </S_div_container>
    </>
  );
};

Article.propTypes = {
  article: PropTypes.object,
};

export default Article;
