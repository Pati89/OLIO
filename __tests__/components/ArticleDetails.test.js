import ArticleDetails from "../../components/ArticleDetails";
import { render } from "../../helpers/test-utils";

const props = {
  article: {
    images: [
      {
        files: {
          large:
            "https://cdn.olioex.com/uploads/photo/file/Y9d--jZv7f0GoLDn3SbH6A/large_image.jpg",
        },
      },
    ],
    title: "test title",
    user: {
      first_name: "Olio user",
    },
  },
};

describe("ArticleDetails", () => {
  it("should render a ArticleDetails", () => {
    const { getByTestId } = render(<ArticleDetails {...props} />);
    expect(getByTestId("articleDetails")).toBeTruthy();
  });

  it("should not render a ArticleDetails", () => {
    const { queryByTestId } = render(<ArticleDetails />);
    expect(queryByTestId("article")).toBeFalsy();
  });

  it("should render a title", () => {
    const { getByText } = render(<ArticleDetails {...props} />);
    expect(getByText("test title")).toBeTruthy();
  });

  it("should render user", () => {
    const { getByText } = render(<ArticleDetails {...props} />);
    expect(getByText("Olio user")).toBeTruthy();
  });

  it("should not render a avatar", () => {
    const { queryByTestId } = render(<ArticleDetails {...props} />);
    expect(queryByTestId("avatar")).toBeFalsy();
  });

  it("should render a avatar", () => {
    const { queryByTestId } = render(
      <ArticleDetails
        article={{
          ...props.article,
          user: {
            current_avatar: {
              small:
                "https://cdn.olioex.com/uploads/avatar/file/oZq8DF3dzLEi3Fnf4XxMrg/small_image.jpg",
            },
          },
        }}
      />
    );
    expect(queryByTestId("avatar")).toBeTruthy();
  });

  it("should render active status", () => {
    const { getByText, getByTestId } = render(
      <ArticleDetails
        article={{
          ...props.article,
          status: "active",
        }}
      />
    );
    expect(getByText("active")).toBeTruthy();
    expect(getByTestId("status")).toHaveStyleRule("background-color", "green");
  });

  it("should render other status", () => {
    const { getByText, getByTestId } = render(
      <ArticleDetails
        article={{
          ...props.article,
          status: "inactive",
        }}
      />
    );
    expect(getByText("inactive")).toBeTruthy();
    expect(getByTestId("status")).toHaveStyleRule("background-color", "grey");
  });

  it("should render description", () => {
    const { getByText, getByTestId } = render(
      <ArticleDetails
        article={{
          ...props.article,
          description:
            "7 X T0714 (yellow)\n4 X T0711 (black)\n1 X E712 (cyan)\n4 X E713 / T0713 (magenta)\n\n",
        }}
      />
    );
    expect(getByTestId("description")).toBeTruthy();
    expect(
      getByText(
        "7 X T0714 (yellow) 4 X T0711 (black) 1 X E712 (cyan) 4 X E713 / T0713 (magenta)"
      )
    ).toBeTruthy();
  });

  it("should render no description message", () => {
    const { getByText, getByTestId } = render(<ArticleDetails {...props} />);
    expect(getByTestId("description")).toBeTruthy();
    expect(getByText("No description")).toBeTruthy();
  });

  it("should render a ArticleDetails match snapshot", () => {
    const { container } = render(<ArticleDetails {...props} />);
    expect(container).toMatchSnapshot();
  });
});
