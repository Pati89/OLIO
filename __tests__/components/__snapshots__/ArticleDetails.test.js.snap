// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ArticleDetails should render a ArticleDetails match snapshot 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 70%;
  background-color: rgb(240, 238, 238);
  padding: 10px 0;
}

@media only screen and (min-width: 640px) {
  .emotion-0 {
    width: 80%;
  }
}

.emotion-1 {
  border: solid black;
  border-width: 0 2px 2px 0;
  display: inline-block;
  padding: 3px;
  -webkit-transform: rotate(135deg);
  -moz-transform: rotate(135deg);
  -ms-transform: rotate(135deg);
  transform: rotate(135deg);
  margin-right: 10px;
}

.emotion-2 {
  width: 70%;
  height: 100%;
  text-align: left;
  color: inherit;
  -webkit-text-decoration: none;
  text-decoration: none;
  overflow: hidden;
  cursor: pointer;
  background-color: white;
}

@media only screen and (min-width: 640px) {
  .emotion-2 {
    width: 80%;
  }
}

.emotion-3 {
  background-color: lightgray;
  position: relative;
  width: 100%;
  height: 400px;
  max-height: 400px;
}

@media only screen and (min-width: 640px) {
  .emotion-3 {
    height: 600px;
    max-height: 600px;
  }
}

.emotion-4 {
  width: 100%;
  padding: 20px;
}

@media only screen and (min-width: 640px) {
  .emotion-4 {
    width: auto;
  }
}

.emotion-5 {
  color: inherit;
  font-weight: bold;
  margin-top: 0;
}

.emotion-6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  justify-content: space-between;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.emotion-7 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.emotion-8 {
  margin-left: 0;
}

.emotion-9 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background-color: grey;
  color: white;
  padding: 5px;
  border-radius: 10px;
  font-size: 0.75rem;
}

.emotion-10 {
  width: calc(100% - 40px);
  margin: 0 20px;
  height: 1px;
  border-top: 1px solid lightgrey;
}

.emotion-11 {
  padding: 0 20px 20px;
}

.emotion-12 {
  font-weight: bold;
}

.emotion-13 {
  white-space: pre-wrap;
}

<div>
  <a
    class="emotion-0"
    href="/"
  >
    <i
      class="emotion-1"
    />
     Back
  </a>
  <div
    class="emotion-2"
    data-testid="articleDetails"
  >
    <div
      class="emotion-3"
    >
      <span
        style="box-sizing: border-box; display: block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px;"
      >
        <img
          alt="test title"
          class="image"
          data-nimg="fill"
          decoding="async"
          src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
          style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
        />
        <noscript />
      </span>
    </div>
    <div
      class="emotion-4"
    >
      <h1
        class="emotion-5"
      >
        test title
      </h1>
      <div
        class="emotion-6"
      >
        <div
          class="emotion-7"
        >
          <span
            class="emotion-8"
          >
            Olio user
          </span>
        </div>
        <div
          class="emotion-9"
          data-testid="status"
        />
      </div>
    </div>
    <div
      class="emotion-10"
    />
    <div
      class="emotion-11"
    >
      <p
        class="emotion-12"
      >
        Description:
      </p>
      <div
        class="emotion-13"
        data-testid="description"
      >
        No description
      </div>
    </div>
  </div>
</div>
`;
